Sources/AsyncHTTPClient/ConnectionPool/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,143 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+final class HTTP1ProxyConnectHandler: ChannelDuplexHandler, RemovableChannelHandler {
+    typealias OutboundIn = Never
+    typealias OutboundOut = HTTPClientRequestPart
+    typealias InboundIn = HTTPClientResponsePart
+
+    enum State {
+        case initialized(EventLoopPromise<Void>)
+        case connectSent(EventLoopPromise<Void>)
+        case headReceived(EventLoopPromise<Void>)
+        case failed(Error)
+        case completed
+    }
+
+    private var state: State
+
+    let targetHost: String
+    let targetPort: Int
+    let proxyAuthorization: HTTPClient.Authorization?
+
+    init(targetHost: String,
+         targetPort: Int,
+         proxyAuthorization: HTTPClient.Authorization?,
+         connectPromise: EventLoopPromise<Void>) {
+        self.targetHost = targetHost
+        self.targetPort = targetPort
+        self.proxyAuthorization = proxyAuthorization
+
+        self.state = .initialized(connectPromise)
+    }
+
+    func handlerAdded(context: ChannelHandlerContext) {
+        precondition(context.channel.isActive, "Expected to be added to an active channel")
+
+        self.sendConnect(context: context)
+    }
+
+    func handlerRemoved(context: ChannelHandlerContext) {
+        switch self.state {
+        case .failed, .completed:
+            break
+        case .initialized, .connectSent, .headReceived:
+            preconditionFailure("Removing the handler, while connecting seems wrong")
+        }
+    }
+
+    func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        preconditionFailure("We don't support outgoing traffic during HTTP Proxy update.")
```

Comment:
just want to make sure - clients will not be able to trigger this precondition no matter what configuration they use?

Reply:
was it copied form the old handler? if yes, then ignore my comment

Reply:
Would it not be a good idea to buffer the data?

Reply:
> just want to make sure - clients will not be able to trigger this precondition no matter what configuration they use?

Yes. We have basically two modes for a channel. Setup and Running. The channel is only returned to the user once it is running. (meaning 100% setup). If someone writes to the `ProxyChannelHandler` this is def. wrong :)

Reply:
> Would it not be a good idea to buffer the data?

@Davidde94 We would develop a feature that we don't use in this case. If we were to put this into `NIOExtras`, I would agree. In AHC it would be purely dead code.